# Exploring Hacker News Posts

## I. Introduction

Hacker News is a site started by the startup incubator [Y Combinator](https://www.ycombinator.com/), where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.  

You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. 

We're specifically interested in posts whose titles begin with either *Ask HN or Show HN*. Users submit *Ask HN* posts to ask the Hacker News community a specific question. Below are a couple examples:  

Ask HN: How to improve my personal website?  
Ask HN: Am I the only one outraged by Twitter shutting down share counts?  
Ask HN: Aby recent changes to CSS that broke mobile?  

Likewise, users submit *Show HN* posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:  

Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'  
Show HN: Something pointless I made  
Show HN: Shanhu.io, a programming playground powered by e8vm  

We'll compare these two types of posts to determine the following:

  * Do Ask HN or Show HN receive more comments on average?
  * Do posts created at a certain time receive more comments on average?

## II. Data Description

Below are descriptions of the columns:  

  * *id*: The unique identifier from Hacker News for the post
  * *title*: The title of the post
  * *url*: The URL that the posts links to, if it the post has a URL
  * *num_points*: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
  * *num_comments*: The number of comments that were made on the post
  * *author*: The username of the person who submitted the post
  * *created_at*: The date and time at which the post was submitted

Here are the first rows of the data set:

In [4]:
opened_file = open('hacker_news.csv')
from csv import reader
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5], '/n')

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']] /n


## III. Methodology

### A. Extracting Ask HN and Show HN Posts

Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

To find the posts that begin with either Ask HN or Show HN, we'll use the string method startswith. Given a string object, say, string1, we can check if starts with, say, dq by inspecting the output of the object string1.startswith('dq'). If string1 starts with dq, it will return True, otherwise it will return False.

In [5]:
print('dataquest'.startswith('Data'))
print('dataquest'.startswith('data'))

False
True


In the example above, the first print call gives us False because dataquest does not start with Data. The second print call prints True because dataquest does start with data. Capitalization matters.

If we wish to control for case, we can use the lower method which returns a lowercase version of the starting string. Here's an example:

In [6]:
print('Dataquest'.lower())

dataquest


Let's use these methods to separate posts beginning with Ask HN and Show HN (and case variations) into two different lists next.

In [29]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts)) 
print(len(other_posts))

1744
1162
17194


There were 1,744 ask posts, 1,162 show posts, and 17,194 other posts.

Below are first five rows of ask_posts list of lists:

In [31]:
print(ask_posts[:5], '\n')

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']] 



Below are first five rows in the show_posts list of lists:

In [32]:
print(show_posts[:5], '\n')

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']] 



### B. Calculating the Average Number of Comments for Ask HN and Show HN Posts

Now that we had separated the ask posts from show posts we will get the average of each posts.

In [36]:
total_ask_comments = 0
for row in ask_posts:
    comments = int(row[4])
    total_ask_comments = total_ask_comments + comments
    avg_ask_comments = total_ask_comments / len(ask_posts)

print(avg_ask_comments)

14.038417431192661


In [37]:
total_show_comments = 0
for row in show_posts:
    comments = int(row[4])
    total_show_comments = total_show_comments + comments
    avg_show_comments = total_show_comments / len(show_posts)

print(avg_show_comments)

10.31669535283993


Ask posts receive more comments on average by getting 14 comments rather than show comments who then receive 10 average comments which is true, because ask posts are more likely to receive more comments. We will focus our analysis on this posts.

### C. Finding the Amount of Ask Posts and Comments by Hour Created

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
2. Calculate the average number of comments ask posts receive by hour created.

In [58]:
import datetime as dt
result_list = []
for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )
    
print(result_list[:5])

counts_by_hour = {}
comments_by_hour = {}
date_format = '%m/%d/%Y %H:%M'

for row in result_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date, date_format).strftime('%H') 
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment

comments_by_hour

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

### D. Calculating the Average Number of Comments for Ask HN Posts by Hour

In [60]:
avg_by_hour = []
for hour in comments_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])

avg_by_hour

[['15', 38.5948275862069],
 ['00', 8.127272727272727],
 ['07', 7.852941176470588],
 ['02', 23.810344827586206],
 ['05', 10.08695652173913],
 ['06', 9.022727272727273],
 ['16', 16.796296296296298],
 ['21', 16.009174311926607],
 ['22', 6.746478873239437],
 ['12', 9.41095890410959],
 ['20', 21.525],
 ['19', 10.8],
 ['14', 13.233644859813085],
 ['23', 7.985294117647059],
 ['13', 14.741176470588234],
 ['17', 11.46],
 ['09', 5.5777777777777775],
 ['18', 13.20183486238532],
 ['01', 11.383333333333333],
 ['11', 11.051724137931034],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['10', 13.440677966101696],
 ['03', 7.796296296296297]]

### E. Sorting and Printing Values from a List of Lists

In [68]:
swap_avg_by_hour = []
for hr in avg_by_hour:
    swap_avg_by_hour.append(
    [hr[1],hr[0]]
    )
swap_avg_by_hour

sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [72]:

print('Top 5 Hours for Asks Posts Comments')
for avg,hr in sorted_swap[:5]:
    print('{}: {:.2f} average comments per post.'.format(
        dt.datetime.strptime(hr,'%H').strftime('%H:%M'),avg))

Top 5 Hours for Asks Posts Comments
15:00: 38.59 average comments per post.
02:00: 23.81 average comments per post.
20:00: 21.52 average comments per post.
16:00: 16.80 average comments per post.
21:00: 16.01 average comments per post.


Based on the above data, the best time to post a question is 3 pm which has an average of 39 comments per posts. 

## IV. Discussion and Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend to post an ask post between 15:00 and 2:00 (3:00 pm est - 2:00 am est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 2:00 (3:00 pm est - 2:00 am est) received the most comments on average.